client/src/transmitter/cast/webrtc.ts

```diff
@@ -0,0 +1,71 @@
+const log = (...data) => console.log('webrtc', ...data)
+
+async function iceHandshake(
+  streamKey: string,
+  localDesc: RTCSessionDescription
+) {
+  const qsKey = streamKey.indexOf('://') > 0 ? 'rtmp' : 'streamKey'
+  const answer = await fetch(
+    `https://webrtmp.justcast.it/webrtc/offer?${qsKey}=${streamKey}`,
+    {
+      method: 'POST',
+      body: JSON.stringify(localDesc),
+      headers: {
+        ['content-type']: 'application/json',
+      },
+    }
+  )
+  if (answer.status !== 200) {
+    throw new Error(`Error response from server: ${answer.status}`)
+  }
+  return (await answer.json()) as RTCSessionDescriptionInit
+}
+
+function castViaWebRTC(stream: MediaStream, streamKey: string): CastSession {
+  const pc = new RTCPeerConnection({
+    iceServers: [
+      {
+        urls: 'stun:stun.l.google.com:19302',
+      },
+    ],
+  })
+  const cast: CastSession = { stop: () => pc.close() }
+
+  stream.getTracks().forEach((track) => pc.addTrack(track, stream))
+
+  pc.oniceconnectionstatechange = () =>
+    log('ice connection state', pc.iceConnectionState)
+  pc.onicegatheringstatechange = () => {
+    if (pc.iceGatheringState !== 'complete') {
+      return
+    }
+    iceHandshake(streamKey, pc.localDescription)
+      .then(async (remoteDesc) => {
+        await pc.setRemoteDescription(new RTCSessionDescription(remoteDesc))
+      })
+      .catch(log)
```

Comment:
Shouldn't we settle for always using async / await + try / catch?

Reply:
I didn't want to make this function `async` otherwise we wouldn't return the `cast` object synchronously to the caller. I could make like an "inner async" function after creating the `cast` object and have only one `catch` here, at least for log.